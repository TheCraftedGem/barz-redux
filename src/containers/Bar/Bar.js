import React, { Component } from "react";
import { connect } from "react-redux";
import { getRhymes, updateLyrics, addBar, setLyrics } from "../../actions";
import { fetchRhymes } from "../../apiCalls";
import { bindActionCreators } from "redux";
import "./Bar.scss";

export class Bar extends Component {
  state = {
    text: "",
    id: 0
  };

  componentDidMount() {
    this.setState({ text: this.props.text })
  }

  handleChange = e => {
    this.setState({ text: e.target.value });
    this.props.updateLyrics(this.state.id, this.state.text);
    localStorage.setItem('lyrics', JSON.stringify(this.props.lyrics));
  };

  handleSelect = async () => {
    const selection = window.getSelection().toString();
    const word = await fetchRhymes(selection);
    this.props.getRhymes(word);
  };

  render() {
    return (
      <article className="bar-wrapper">
        <p className="bar-number">{this.props.number}</p>
        <input
          className="bar-input"
          id={this.state.id}
          type="text"
          name="text"
          value={this.state.text}
          onChange={this.handleChange}
          onSelect={this.handleSelect}
        />
        <button className="bar-close">X</button>
      </article>
    );
  }
}

export const mapStateToProps = state => ({
  lyrics: state.lyrics,
  rhymes: state.rhymes
});

export const mapDispatchToProps = dispatch =>
  bindActionCreators({ getRhymes, updateLyrics, addBar, setLyrics }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Bar);
