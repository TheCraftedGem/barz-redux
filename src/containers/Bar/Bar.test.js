import React from "react";
import { Bar, mapStateToProps, mapDispatchToProps } from "./Bar";
import {
  getRhymes,
  updateLyrics,
  addBar,
  setLyrics,
  updateBarActive,
  deleteBar
} from "../../actions";
import { shallow } from "enzyme";

describe("Bar", () => {
  let wrapper;
  let instance;

  beforeEach(() => {
    wrapper = shallow(<Bar text="" id={0} active={false} />);
    instance = wrapper.instance();
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have default state", () => {
    expect(instance.state).toEqual({ text: "", id: 0, active: false });
  });

  describe("mapStateToProps", () => {
    it("should return an array of lyrics objects", () => {
      const mockState = {
        lyrics: [
          {
            title: "mockLyric",
            id: 1,
            date: "11/25/1989",
            active: true,
            bars: [{ id: 1, text: "mockBar" }]
          }
        ],
        rhymes: [{ word: "mockWord" }]
      };
      const expected = {
        lyrics: [
          {
            title: "mockLyric",
            id: 1,
            date: "11/25/1989",
            active: true,
            bars: [{ id: 1, text: "mockBar" }]
          }
        ],
        rhymes: [{ word: "mockWord" }]
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe("mapDispatchToProps", () => {
    it("calls dispatch with a updateLyrics action when handleChange is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = updateLyrics(
        {
          barId: 1,
          text: "mockTitle",
        }
      );
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateLyrics(
        {
          barId: 1,
          text: "mockTitle",
        }
      );
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it("calls dispatch with a getRhymes action when handleSelect is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = getRhymes([
        { word: "mockWord1" },
        { word: "mockWord2" }
      ]);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.getRhymes([{ word: "mockWord1" }, { word: "mockWord2" }]);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it("calls dispatch with a addBar action when handleKeyDown is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = addBar({
        id: Date.now(),
        text: "",
        active: true
      });
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addBar({
        id: Date.now(),
        text: "",
        active: true
      });
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it("calls dispatch with a updateBarActive action when handleKeyDown is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = updateBarActive({ barId: 1, active: true });
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateBarActive({ barId: 1, active: true });
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it("calls dispatch with a deleteBar action when handleKeyDown is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = deleteBar({ id: 1 });
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.deleteBar({ id: 1 });
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
