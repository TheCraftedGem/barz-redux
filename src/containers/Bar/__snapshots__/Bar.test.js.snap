// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar should match the snapshot 1`] = `ShallowWrapper {}`;
